storage/api/src/main/java/feast/storage/api/retrieval/HistoricalRetrievalResult.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2018-2020 The Feast Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.storage.api.retrieval;
+
+import com.google.auto.value.AutoValue;
+import feast.serving.ServingAPIProto.DataFormat;
+import feast.serving.ServingAPIProto.JobStatus;
+import java.io.Serializable;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Result of a historical feature retrieval request. */
+@AutoValue
+public abstract class HistoricalRetrievalResult implements Serializable {
+
+  public abstract String getId();
+
+  public abstract JobStatus getStatus();
+
+  @Nullable
+  public abstract String getError();
+
+  @Nullable
+  public abstract List<String> getFileUris();
+
+  @Nullable
+  public abstract DataFormat getDataFormat();
+
+  /**
+   * Instantiates a {@link HistoricalRetrievalResult} indicating that the retrieval was a failure,
+   * together with its associated error.
+   *
+   * @param id retrieval id identifying the retrieval request.
+   * @param error error that occurred
+   * @return {@link HistoricalRetrievalResult}
+   */
+  public static HistoricalRetrievalResult errorResult(String id, Exception error) {
```

Comment:
Actually I wonder how it'd look if you dropped `Result` from the class name altogether… Since it's representative of a (possibly unstarted) process to produce results, but doesn't actually directly contain the result data.

Reply:
Actually, it always represents a completed retrieval request. I'll update it to your former suggestion.

Reply:
Okay, in that case it makes sense.

I feel a little reserved though about committing to the pattern of out-of-band file-based result data as API, i.e. `getFileUris()`. Not sure if large data sets will ever be streamed through RPC responses, but either of these options compromise the possibility for user jobs doing retrieval and processing the results to leverage data locality in systems where that could be achieved—see discussion starting at https://github.com/gojek/feast/issues/482#issuecomment-590827523

Reply:
To be clear, I think there's nothing actionable about my prior comment for this PR, I think it's a bigger picture discussion for #567.